File: src/array/SortHashMapValues.java
package array;

import java.util.*;

public class SortHashMapValues {

    public static void main(String[] args) {
        //int[] arr = {5,5,4,6,4};
        int[] arr = {9,9,9,2,5};

        //Storing the frequency of above elements in hashmap..
        HashMap<Integer, Integer> map = storeFrequencyOfElementsInMap(arr);

        //convert hashmap to AL..
        ArrayList<Map.Entry<Integer, Integer>> arrayList = new ArrayList<>(map.entrySet());

        //Sort the AL in descending order w.r.t values. So, implementing Comparator..
        Collections.sort(arrayList, (o1, o2) -> o2.getValue().compareTo(o1.getValue()));

        //Store the sorted AL's elements in Linked HashMap..
        LinkedHashMap<Integer, Integer> linkedHashMap = new LinkedHashMap<>();
        for (Map.Entry<Integer, Integer> entry: arrayList) {
            linkedHashMap.put(entry.getKey(), entry.getValue());
        }

        //print linkedHashMap..
        for (Map.Entry<Integer, Integer> entry: linkedHashMap.entrySet()) {
            System.out.println(entry.getKey()+":"+entry.getValue());
        }

        //Output of this problem
        int currentFreq = 0;
        int prevElement = arr[0];
        ArrayList<Integer> output = new ArrayList<>();
        for (int index = 0; index < arr.length; index++) {
            int freq = map.get(arr[index]);
            if (freq >= currentFreq && arr[index] <= prevElement) {
                output.add(0, arr[index]);
            } else {
                output.add(output.size() , arr[index]);
            }
            currentFreq = freq;
            prevElement = arr[index];
        }

        //Iterate output..
        for (int i = 0; i < output.size(); i++) {
            System.out.print(output.get(i)+" ");
        }
    }

    private static HashMap<Integer,Integer> storeFrequencyOfElementsInMap(int[] array) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < array.length; i++) {
            if (map.containsKey(array[i])) {
                map.put(array[i], map.get(array[i]) + 1);
            } else {
                map.put(array[i], 1);
            }
        }
        return map;
    }
}

File: src/array/SortHashMapValues.kt
/*
package array

import java.security.KeyStore
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.LinkedHashMap

fun main() {

    //Sorting Elements of an Array by Frequency..
    val arr = intArrayOf(5, 5, 4, 6, 4)
    val hashMap = storeFrequencyOfElementsInMap(arr)
    for (entry in hashMap.entries) {
        println("${entry.key} : ${entry.value}")
    }

    //Sort the hashmap w.r.t values in descending order..
    val entries = hashMap.entries
    val arrayList = ArrayList(entries)

    arrayList.sortWith(Comparator { o1, o2 -> o2.value.compareTo(o1.value) })

    //insert in linkedHashMap..
    val linkedHashMap = LinkedHashMap<Int, Int>()

    val entry = java.util.Map.Entry<Int, Int>

    for (entry: Map.Entry<Int, Int> in arrayList.indices)

}

fun storeFrequencyOfElementsInMap(array: IntArray) : HashMap<Int, Int> {
    val hashMap = HashMap<Int, Int>()
    for (index in array.indices) {
        if (hashMap.containsKey(array[index])) {
            val value = hashMap[array[index]]
            value?.plus(1)?.let { hashMap.put(array[index], it) }
        } else {
            hashMap[array[index]] = 1
        }
    }
    return hashMap
}*/
